### PR TITLE
File file endpoints and unit tests

### PR DESCRIPTION
**Pull request by carycheng**

### CODE REVIEW

**Comment by boxcla** [comment]
Hi @carycheng, thanks for the pull request. Before we can merge it, we need you to sign our Contributor License Agreement. You can do so electronically here: http://opensource.box.com/cla

Once you have signed, just add a comment to this pull request saying, "CLA signed". Thanks!

**Comment by anthonywee** [comment]
commit title typo "File file"

File: boxsdk/client/client.py [hunk 1]
```diff
@@ -391,3 +394,158 @@ def get_url(self, endpoint, *args):
         """
         # pylint:disable=no-self-use
         return self._session.get_url(endpoint, *args)
+
+
```
Comment (anthonywee) [inline boxsdk/client/client.py]: white space

File: boxsdk/client/client.py [hunk 2]
```diff
@@ -391,3 +394,158 @@ def get_url(self, endpoint, *args):
...
+
+    def retention_policy(self, retention_id):
+        """
+        Initialize a :class: `RetentionPolicy` object, whose box id is retention_id.
```
Comment (anthonywee) [inline boxsdk/client/client.py]: extra space in front of :class:

File: boxsdk/object/file.py [hunk 3]
```diff
@@ -71,6 +71,43 @@ def download_to(self, writeable_stream):
         for chunk in box_response.network_response.response_as_stream.stream(decode_content=True):
             writeable_stream.write(chunk)
 
+
+    def download_url(self, version_id=None, range=None, box_api=None):
```
Comment (anthonywee) [inline boxsdk/object/file.py]: would be good to rename `range` since that is a built in as well

File: boxsdk/object/file.py [hunk 4]
```diff
@@ -289,3 +326,90 @@ def get_shared_link_download_url(
...
+            params['max_height'] = max_height
+        if min_width is not None:
+            params['min_width'] = min_width
+        if max_width is not None:
+            params['max_width'] = max_width
+        box_response = self._session.get(url, params=params)
+        for chunk in box_response.network_response.response_as_stream.stream(decode_content=True):
+            writeable_stream.write(chunk)
+        # response = box_response.json()
+        # return self.__class__(
```
Comment (anthonywee) [inline boxsdk/object/file.py]: remove commented lines

File: boxsdk/object/file.py [hunk 5]
```diff
@@ -289,3 +326,90 @@ def get_shared_link_download_url(
...
+            params['min_height'] = min_height
+        if max_height is not None:
+            params['max_height'] = max_height
+        if min_width is not None:
+            params['min_width'] = min_width
+        if max_width is not None:
+            params['max_width'] = max_width
+        box_response = self._session.get(url, params=params)
+        for chunk in box_response.network_response.response_as_stream.stream(decode_content=True):
+            writeable_stream.write(chunk)
```
Comment (anthonywee) [inline boxsdk/object/file.py]: where is `writeable_stream` defined?

File: boxsdk/object/file.py [hunk 6]
```diff
@@ -71,6 +71,43 @@ def download_to(self, writeable_stream):
...
+        :type version_id:
+            `unicode` or None
+        :param range:
+            The value of range in bytes. Format should be `bytes={start_range}-{end_range}`
+        :type range:
+            `unicode` or None
+        :param box_api:
+            Set to `shared_link=SHARED_LINK_URL` or `shared_link=SHARED_LINK_URL&shared_link_password=PASSWORD`
+        :type: box_api:
+            `unicode` or None
```
Comment (anthonywee) [inline boxsdk/object/file.py]: :return:

File: boxsdk/object/item.py [hunk 7]
```diff
@@ -322,3 +322,81 @@ def delete(self, params=None, etag=None):
...
+        :rtype:
+            :class:`BoxObjectCollection`
+        """
+        return MarkerBasedObjectCollection(
+            session=self._session,
+            url=self.get_url('metadata'),
+            limit=500,
+            marker=None,
+            fields=None,
+            return_full_pages=False
```
Comment (anthonywee) [inline boxsdk/object/item.py]: trailing comma

File: boxsdk/object/item.py [hunk 8]
```diff
@@ -322,3 +322,81 @@ def delete(self, params=None, etag=None):
...
+            Key value pair to add to metadata.
+        :type key_1:
+            `unicode`
+        :param scope:
+            The scope of the metadata object.
+        :type scope:
+            `unicode`
+        :param template:
+            The key of the template
+        :type template:
```
Comment (anthonywee) [inline boxsdk/object/item.py]: update docs for key_2

File: boxsdk/object/item.py [hunk 9]
```diff
@@ -322,3 +322,81 @@ def delete(self, params=None, etag=None):
...
+            `unicode`
+        """
+        url = self.get_url('metadata', scope, template)
+        body = {
+            key_1
```
Comment (anthonywee) [inline boxsdk/object/item.py]: perhaps doesn't need to be multi line?

File: boxsdk/object/item.py [hunk 10]
```diff
@@ -322,3 +322,81 @@ def delete(self, params=None, etag=None):
...
+
+
+    def create_metadata(self, key_1, scope, template, key_2=None):
+        """
+        Create a metadata instance for a file.
+
+        :param key_1:
+            Key value pair to add to metadata.
+        :type key_1:
+            `unicode`
```
Comment (anthonywee) [inline boxsdk/object/item.py]: is this supposed to be `dict`?

File: boxsdk/object/item.py [hunk 11]
```diff
@@ -322,3 +322,81 @@ def delete(self, params=None, etag=None):
...
+            The scope of the metadata object.
+        :type scope:
+            `unicode`
+        :param template:
+            The key of the template
+        :type template:
+            `unicode`
+        """
+        url = self.get_url('metadata', scope, template)
+        response = self._session.delete(url, data=json.dumps(body), headers=headers).json()
```
Comment (anthonywee) [inline boxsdk/object/item.py]: body not defined?

**Pull request closed without merging**